pyamg/aggregation/smooth.py

```diff
@@ -187,28 +204,44 @@ def jacobi_prolongation_smoother(S, T, C, B, omega=4.0/3.0, degree=1,
         # apply satisfy constraints so that U*B = 0
         P = T
         for i in range(degree):
-            U = (D_inv_S*P).tobsr(blocksize=P.blocksize)
+            if sparse.isspmatrix_bsr(P):
+                U = (D_inv_S*P).tobsr(blocksize=P.blocksize)
+                U_block = U.blocksize
+            else:
+                U = D_inv_S*P
+                U_block = [1,1]
+
+            cost[0] += P.nnz / float(S.nnz)
 
-            # Enforce U*B = 0 (1) Construct array of inv(Bi'Bi), where Bi is B
+            # (1) Enforce U*B = 0. Construct array of inv(Bi'Bi), where Bi is B
             # restricted to row i's sparsity pattern in Sparsity Pattern. This
             # array is used multiple times in Satisfy_Constraints(...).
             BtBinv = compute_BtBinv(B, U)
```

Comment:
The reason I don't pass cost into functions, e.g. compute_BtBinv, is because the cost is all tracked in WUs on the current grid, which means it is respective to the number of nonzeros in A. Many of these functions do not take in A as a parameter, which means I either need to also pass in A.nnz, pass in a temp variable to get FLOPs, then let cost += temp/A.nnz, or track global complexity only in FLOPs. I didn't like passing in A.nnz because it's extra parameters; didn't like the counting FLOPs via cost[] in a function, as it would be inconsistent with the rest of the functions that count WUs in cost[]; and didn't like the complexity counting all in FLOPs instead of WUs partially due to overflow concern, and because something about the whole thing made more sense in WUs to me.

So, I'm happy to discuss modifications, but, it's not quite as simple as just adding cost[] to the extra functions.